tree/ntuple/v7/inc/ROOT/RPageStorage.hxx

```diff
@@ -420,6 +420,30 @@ protected:
    /// `RPageAllocatorHeap`; use `RPageAllocatorHeap::DeletePage()` to deallocate returned pages.
    RPage UnsealPage(const RSealedPage &sealedPage, const RColumnElementBase &element, DescriptorId_t physicalColumnId);
 
+   /// Prepare a page range read for the column set in `clusterKey`.  Specifically, pages referencing the
+   /// `kTypePageZero` locator are filled in `pageZeroMap`; otherwise, `perPageFunc` is called for each page. This is
+   /// commonly used as part of `LoadClusters()` in derived classes.
+   template <typename PerPageFuncT>
```

Comment:
I'd prefer avoiding the template and using an `std::function` parameter instead. That would allow us to move this from header to compilation unit, and it would clarify the parameter types of `perPageFunc`.

Reply:
This was a concious decision; unfortunately, using a `std::function` would make the call to the lambda non-inlineable, resulting in an indirect call for each page in the range.  For small functions, e.g. in RPageStorageFile.cxx (see [here](https://github.com/root-project/root/pull/13388/files#diff-311b1306fdf38fbec1f6e0eeb3065c34b411f6d3f6f05614eb75cc02968f2a84R444-R446)), the relative cost of the call increases.
(See also related interesting read: https://vittorioromeo.info/index/blog/passing_functions_to_functions.html)

What about leaving here only the declaration and moving the definition out-of-class at the end of RPageStorage.hxx?  I think it makes it more readable; for the parameter types of `perPageFunc`, we could document above.  What do you think?

Reply:
Ok, I understand that the call cannot be inlined. But cannot asses yet if that is a problem in this particular case. Can you collect numbers that show the performance effect.

Reply:
Apparently, in a synthetic test reading 8 `std::uint32_t` fields and 10M entries, the cost of not inlining is negligible (~ a few $us$), especially compared to the cost of the `ReadV()` call that follows.

In light of this, I effectively moved the implementation to the .cxx file :+1:.